scripts/igmp_query_injector.py

```diff
@@ -274,134 +69,68 @@ def inject(self):
+                    self.inject_to_vif(vif)
+                watcher.unwatch(path, vif)
 
-        t = threading.Thread(target=watch.process_all_watches)
+    def _inject_with_connection_state_check(self):
+        watcher = XSWatcher()
+        for vif in self.vifs:
+            state_path, backend_state_path = get_vif_state_path(vif)
+            # watch both frontend and backend vif state
+            watcher.watch(state_path, vif)
+            watcher.watch(backend_state_path, vif)
+
+        # We have 2 options to handle the blocking function `xs.read_watch`
+        # 1. Single thread with Unix alarm signal
+        #    The signal handler cannot be properly invoked because when signal received the process context is in
+        #    C extension but not python interpreter, so the interpreter cannot call the signal handler.
+        #    So we should not apply this option.
+        # 2. Multi thread with threading.event or timeout associated join
+        #    Create a new thread to invoke the blocking methon `xs.read_watch`. In the main thread using a timeout
+        #    associated join method for waiting. We will apply this option.
+        t = threading.Thread(target=self.check_and_inject, args=(watcher,))
         t.daemon = True
         t.start()
-        remain_time = self.vif_connected_timeout
-
-        while remain_time > 0:
-            start_time = time.time()
-            try:
-                task = watch.get_task(remain_time)
-                vif = task.vif
-                self.inject_to_vif(vif.vif_name, vif.mac)
-                vif.mark_injected()
-                if task.is_last:
-                    return
-            except Queue.Empty as e:
-                # timeout
-                break
-
-            remain_time = remain_time - (time.time() - start_time)
-
-        for vif in vifs:
-            if not vif.injected:
-                log.warning("Value of '%s' does not change to '%s' in %d seconds." %
-                            (vif.state_path, expected_vif_state, self.vif_connected_timeout))
-                log.warning("Don't inject IGMP query to vif: %s, mac: %s" % (vif.vif_name, vif.mac))
+        t.join(self.vif_connected_timeout)
+        if watcher.watches:
+            log.warning('Wait vif state change timeout')
+            for vif in watcher.watches.itervalues():
+                log.warning("Vif:%s state did not change to '%s', don't inject IGMP query to mac: %s" %
+                            (vif, VIF_CONNECTED_STATE, get_vif_mac(vif)))
 
     def _inject_without_connection_state_check(self):
-        self.inject_to_vifs(self.vifs)
-
+        for vif in self.vifs:
+            self.inject_to_vif(vif)
 
-class BridgeInjector(IGMPQueryInjector):
-    RE_VIF = re.compile(r'^vif\d+\.\d+$')
 
-    def __init__(self, max_resp_time, bridges):
-        super(BridgeInjector, self).__init__(max_resp_time)
-        self.bridges = bridges
+def domid_vifid_of_vif(vif):
```

Comment:
I think it would be neater to have this function return the vif path, then you can just append the key you want to that path:
```
def vif_path(vif):
   domid, vifid = vif.split('vif')[1].split('.')
   dompath = g_xs_handler.get_domain_path(domid)
   return os.path.join(os.path.join(dompath, "device/vif", vifid))

def get_vif_mac(vif):
   return g_xs_handler.read("", os.path.join(vif_path(vif), "mac"))
```
You would do something similar for the backend paths.